### PR TITLE
WIP Saying good bye to *_where_sql() functions

### PR DESCRIPTION
**Pull request by hypeJunction**
Needs #11403

- [x] Add `River` repository tests
- [x] Add `elgg_get_tags()` tests

I am not going to bother with search tests on this. I am going to revive #10688 and add tests there.

### CODE REVIEW

File: engine/classes/Elgg/Database/EntityTable.php [hunk 1]
```diff
@@ -707,63 +546,20 @@ public function getEntityTimeWhereSql($table, $time_created_upper = null,
...
-
-		$where[] = _elgg_get_access_where_sql(['table_alias' => '']);
-
-		$sql = "SELECT DISTINCT EXTRACT(YEAR_MONTH FROM FROM_UNIXTIME(time_created)) AS yearmonth
-			FROM {$this->db->prefix}entities where ";
-
-		foreach ($where as $w) {
-			$sql .= " $w and ";
-		}
+		$options = [
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: Need a test here. Should we move this to the blog plugin? I have never used it.

File: engine/classes/Elgg/Database/River.php [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+	 * Make sure all three are accessible by the user
+	 *
+	 * @param QueryBuilder $qb Query builder
+	 *
+	 * @return CompositeExpression|mixed|null|string
+	 */
+	public function buildEntityClauses($qb) {
+		$ands = [];
+
+		$qb->joinEntitiesTable('rv', 'subject_guid', 'inner', 'se');
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/River.php]: Check if river API allows non-users to be subjects. If not, then we can remove this

File: engine/lib/tags.php [hunk 3]
```diff
@@ -74,120 +74,40 @@ function string_to_tag_array($string) {
...
+		->orderBy('total', 'desc');
 
-	$limit = sanitise_int($options['limit']);
-	$query .= " LIMIT {$limit} ";
+	$options = new \Elgg\Database\QueryOptions($options);
+	$alias = $qb->joinEntitiesTable('md', 'entity_guid', 'inner', 'e');
+	$qb->addClause(\Elgg\Database\Clauses\EntityWhereClause::factory($options), $alias);
 
-	return get_data($query);
+	return \Elgg\Database\Metadata::find($options);
```
Comment (hypeJunction) [inline engine/lib/tags.php]: Go back to `get_data`

**Comment by jdalsem** [comment]
let me know when this is ready for review

**Comment by hypeJunction** [comment]
Not yet, more stuff coming. Hope to get it out of my life later today or tomorrow. Need to get back to real world problems

**Comment by hypeJunction** [comment]
I think I am done here. Should have split this up into multiple PRs, but it was like a roller coaster. 

I have added some performance improvements to metadata and annotations, as I have noticed there were a ton of useless queries happening during testing. Notably, metadata cache now cacheds IDs, so we don't have to read them from the database every time metadata is updated or created.

**Comment by hypeJunction** [comment]
This also halves the build time. I simplified unit test entity table mock and stoped querying for active plugins if boot cache is populated

**Comment by hypeJunction** [comment]
Could someone wipe the caches on Travis. I think we are running out of disk space.

**Comment by jdalsem** [comment]
> Could someone wipe the caches on Travis. I think we are running out of disk space.

done... why did you think we ran out of it btw?

**Comment by hypeJunction** [comment]
during one of the builds it failed on multiple boxes with memory error

File: docs/appendix/upgrade-notes/2.x-to-3.0.rst [hunk 4]
```diff
@@ -203,6 +206,7 @@ Deprecated APIs
  * ``elgg_list_entities_from_relationship``: Use ``elgg_list_entities()``
  * ``elgg_list_entities_from_private_settings``: Use ``elgg_list_entities()``
  * ``elgg_list_entities_from_access_id``: Use ``elgg_list_entities()``
+ * ``elgg_batch_delete_callback``
```
Comment (jeabakker) [inline docs/appendix/upgrade-notes/2.x-to-3.0.rst]: why remove this function? It can be helpful sometimes

**Review by jeabakker — APPROVED** [review]

File: docs/appendix/upgrade-notes/2.x-to-3.0.rst (hunk 4 above)
Comment (hypeJunction) [reply docs/appendix/upgrade-notes/2.x-to-3.0.rst]: I think it creates a anti-pattern. It makes it harder to read the code when used with ElggBatch. I'd rather we write explicit code with object instances that are type hinted.

**Comment by hypeJunction** [comment]
You may want to review and merge this before it goes out of sync. I don't have any more time to invest into this.

File: engine/tests/phpunit/unit/Elgg/Views/ViewRenderingTestCase.php [hunk 5]
```diff
@@ -74,6 +92,7 @@ public function assertViewOutput($expected, $view, $view_vars = [], $viewtype =
 	 */
 	public function testCanRenderViewWithEmptyVars($view, $viewtype) {
 		if (!elgg_view_exists($view, $viewtype)) {
+			dump($view);
```
Comment (jeabakker) [inline engine/tests/phpunit/unit/Elgg/Views/ViewRenderingTestCase.php]: is this left over debug code?

Comment (hypeJunction) [reply engine/tests/phpunit/unit/Elgg/Views/ViewRenderingTestCase.php]: Oops, yes

Comment (jeabakker) [reply engine/tests/phpunit/unit/Elgg/Views/ViewRenderingTestCase.php]: if you could remove that, i'll merge this

Comment (hypeJunction) [reply engine/tests/phpunit/unit/Elgg/Views/ViewRenderingTestCase.php]: done

**Pull request merged**